### PR TITLE
Iterative 2-Pass Refinement: Self-Conditioning (AlphaFold2 Recycling)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Run the Transolver model TWICE on each input — pass-1 produces a standard prediction, then pass-2 receives the pass-1 output as additional input conditioning and produces a refined prediction. **Same model weights for both passes (no extra parameters).** The loss is computed only on pass-2 output.

**Why this should work:** The model's own first-pass predictions carry information about its uncertainty and systematic biases that is not present in the input geometry alone. In the inter-foil gap region (the primary source of p_tan failures), pass-1 will have structured errors. Pass-2, conditioned on these errors, can learn a correction function `f(geometry, pass1_prediction) → refined_prediction`.

**Prior art:**
- **AlphaFold2 recycling** (3 passes, shared weights, structure→structure): The most successful example of iterative refinement in scientific ML. Recycling improved pLDDT by ~5 points.
- **Diffusion self-conditioning** (Chen et al. 2022, "Analog Bits"): Conditioning generation on the model's own prior estimate improves quality at zero cost during generation.
- **Iterative error correction** in PDE solvers: Classical multigrid methods refine coarse solutions.

**This is a bold architectural change** — it doubles training compute but adds zero parameters.

### Step 1: Add a correction projection to the Transolver

The pass-1 output is 3-dim (Ux, Uy, p) or 4-dim depending on output channels. We need a way to feed this back into the model. Add a small linear layer that projects the pass-1 prediction into the hidden space and ADDS it to the standard input features.

In `Transolver.__init__`, after the main `preprocess` layers:

```python
self.iterative_refinement = cfg.iterative_refinement
self.iterative_warmup_epoch = cfg.iterative_warmup_epoch  # default: 60
if self.iterative_refinement:
    self.correction_proj = nn.Linear(self.out_dim, self.hidden_dim, bias=False)
    nn.init.zeros_(self.correction_proj.weight)  # Zero-init: starts as identity
```

### Step 2: Modify the forward pass

In `Transolver.forward()`, split the forward logic so it can be called twice. The simplest approach: extract the core forward computation into a helper method `_forward_core(fx, ...)` if not already factored. Otherwise, wrap the forward in a 2-pass loop:

```python
def forward(self, x, fx, ..., epoch=0):
    # Standard feature preprocessing (DSDF, normalization, etc.)
    fx = self.preprocess(fx)  # [B, N, H]
    
    if self.iterative_refinement and epoch >= self.iterative_warmup_epoch:
        # Pass 1: standard forward (detached to save memory)
        with torch.no_grad():
            pass1_fx = fx  # save input features
            for block in self.blocks:
                fx = block(fx, ...)
            pass1_pred = self.output_head(fx)  # [B, N, out_dim]
        
        # Pass 2: add correction signal from pass-1 predictions
        fx = pass1_fx  # reset to original features
        correction = self.correction_proj(pass1_pred.detach())  # [B, N, H]
        fx = fx + correction  # condition on pass-1 output
        
        for block in self.blocks:
            fx = block(fx, ...)
        pred = self.output_head(fx)
        return pred  # train on pass-2 output
    else:
        # Standard single-pass forward (before warmup or at inference without refinement)
        for block in self.blocks:
            fx = block(fx, ...)
        return self.output_head(fx)
```

**IMPORTANT implementation notes:**
1. `torch.no_grad()` on pass-1 saves ~50% memory — gradients only flow through pass-2. This is critical to fit in 96GB VRAM.
2. The `correction_proj` is zero-initialized, so at the start of warmup (epoch 60), the correction signal is zero and pass-2 ≡ pass-1 — no discontinuity.
3. Before `iterative_warmup_epoch`, use standard single-pass to avoid wasting compute on garbage pass-1 predictions.
4. The existing `_forward_core` structure in `Transolver.forward` may not be cleanly separable — adapt to the actual code structure. The key is: run the transformer blocks twice, with correction added before the second run.

### Step 3: Handle EMA and validation

At **validation/inference time**, always use 2-pass if `iterative_refinement=True` and epoch >= warmup. The EMA model should also use 2-pass refinement.

### Step 4: Add config flags

```python
# Config dataclass:
iterative_refinement: bool = False
iterative_warmup_epoch: int = 60

# argparse:
parser.add_argument('--iterative_refinement', action='store_true')
parser.add_argument('--iterative_warmup_epoch', type=int, default=60)
```

### Step 5: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/iter-refine-s42" --wandb_group phase6/iterative-refinement \
  --iterative_refinement --iterative_warmup_epoch 60 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73: same flags, --seed 73, wandb_name "thorfinn/iter-refine-s73"
```

### What to report

| Config | Seed | p_in | p_oodc | p_tan | p_re | VRAM | W&B |
|--------|------|------|--------|-------|------|------|-----|
| iterative 2-pass | 42 | | | | | | |
| iterative 2-pass | 73 | | | | | | |
| **iter avg** | — | | | | | | |
| **Baseline** | — | 13.05 | 7.70 | 28.60 | 6.55 | ~46GB | d7l91p0x, j9btfx09 |

Also log:
1. **VRAM usage** — expect ~60-70GB with no_grad pass-1 (vs 46GB baseline). Report if it OOMs.
2. **Training time** — expect ~1.5-1.8x baseline per epoch after warmup.
3. **Validation loss at epoch 59 (pre-warmup) vs epoch 61 (post-warmup)** — should be continuous (no discontinuity from zero-init correction).
4. If the 2-pass approach OOMs, try reducing batch size as a fallback, or report and we'll adjust.

## Baseline

| Metric | Target |
|--------|--------|
| p_in   | < 13.05 |
| p_oodc | < 7.70  |
| **p_tan** | **< 28.60** |
| p_re   | < 6.55  |

W&B baseline: d7l91p0x (s42), j9btfx09 (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### Iterative 2-Pass Refinement (AlphaFold2-style recycling)

| Config | Seed | p_in | p_oodc | p_tan | p_re | VRAM | W&B |
|--------|------|------|--------|-------|------|------|-----|
| iterative 2-pass | 42 | 14.5 | 8.6 | 30.3 | 7.1 | 46.6GB | u0nsat3a |
| iterative 2-pass | 73 | 14.8 | 9.1 | 30.7 | 7.4 | 46.4GB | 5yxjk7be |
| **iter avg** | — | **14.65** | **8.85** | **30.5** | **7.25** | ~46.5GB | — |
| **Baseline** | — | 13.05 | 7.70 | 28.60 | 6.55 | ~46GB | d7l91p0x, j9btfx09 |

**Verdict: All metrics worse than baseline. The hypothesis did not work.**

### Key observations

1. **VRAM usage**: 46.4-46.6GB — essentially identical to baseline (~46GB). The `torch.no_grad()` on pass-1 was highly effective at avoiding the expected VRAM doubling.

2. **Training time**: Epoch time increased from ~71s (single-pass, epochs 1-60) to ~92s (2-pass, epochs 61+), a **1.3x** overhead — much better than the expected 1.5-1.8x. However, this still meant fewer total epochs: 131 epochs vs what baseline achieves in the same 180-minute window.

3. **Warmup transition (epoch 59→61)**: Smooth and continuous, validating the zero-init correction design:
   - s42: val_tandem_transfer 1.108 → 1.098 → 1.098 (essentially no change)
   - s73: val_tandem_transfer 1.140 → 1.095 → 1.091 (continuous)
   - Epoch time: 71s → 165s (first 2-pass epoch had initialization overhead), then settled at ~92s

4. **Effective training epochs**: Only 71 epochs of 2-pass training (epochs 60-131) before the 180-min wall clock hit. The model was still actively converging when training stopped. The 1.3x epoch time cost meant ~20 fewer total epochs vs baseline.

### What happened

The iterative 2-pass refinement **underperformed the baseline on all metrics**, with p_tan being 30.5 vs 28.60 (+6.6% worse). The likely cause is a combination of:

1. **Reduced effective training budget**: The 2-pass approach costs 1.3x per epoch. Over 180 minutes, this translates to ~131 epochs vs baseline's likely ~150+ single-pass epochs. The model was still converging at termination.

2. **Late activation**: The correction signal only kicks in at epoch 60, and with only ~71 2-pass epochs total, the model had limited time to learn the correction function. The correction_proj was zero-initialized and needs many epochs to learn meaningful corrections.

3. **The premise may be flawed for this architecture**: Unlike AlphaFold2 (which has structured outputs with clear geometric constraints), CFD surface predictions may not have systematic pass-1 error patterns that are easy for pass-2 to correct. The inter-foil gap pressure errors may be driven by missing information (flow interactions) rather than correctable biases.

### Commands used

```bash
# Seed 42 (GPU 0)
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py --agent thorfinn \
  --wandb_name "thorfinn/iter-refine-s42" --wandb_group phase6/iterative-refinement \
  --iterative_refinement --iterative_warmup_epoch 60 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73 (GPU 1) — same flags, --seed 73
```

### Suggested follow-ups

1. **Longer training budget**: If the 180-min wall clock could be extended (or the model made faster), the 2-pass approach might close the gap — it was still converging at termination.
2. **Earlier warmup**: Starting the 2-pass at epoch 30 instead of 60 would give more refinement epochs, though pass-1 quality would be worse.
3. **Partial refinement**: Instead of running all transformer blocks twice, run only the last 1-2 blocks in pass-2 — this would reduce the per-epoch overhead while still providing a correction signal.
4. **Gradient through pass-1**: Currently pass-1 is detached. Allowing gradients (at higher VRAM cost) might help the model learn to produce more informative pass-1 outputs.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2165 because ADVISOR: Closing — all metrics regress, p_tan = 30.5 (+6.6% vs 28.60 baseline).

Key findings from the experiment:
1. **VRAM**: torch.no_grad() on pass-1 effectively prevented VRAM doubling (~46.5GB vs ~46GB baseline). Great implementation.
2. **Training budget**: 1.3x epoch time → only 131 total epochs vs baseline's ~150+. Model was still converging at 180-min wall clock — the approach fundamentally trades epochs for refinement quality, and with our fixed time budget, that trade doesn't pay off.
3. **Smooth warmup**: Zero-init correction design validated — no transition artifacts at epoch 60.
4. **Core issue**: The 2-pass premise may be flawed for this architecture. Unlike AlphaFold2 (structured outputs with geometric constraints), CFD surface pressure errors are driven by missing flow interaction information, not correctable biases. Pass-2 can't recover what pass-1 couldn't encode.

Re: your suggested follow-ups — extending training time is outside our constraints (SENPAI_TIMEOUT_MINUTES). Earlier warmup and partial refinement are interesting but the fundamental issue is information-theoretic, not compute-related. Closing this direction.

Dead end: iterative 2-pass refinement. The approach trades training epochs for refinement quality, losing on our fixed-budget constraint. Even with unlimited time, the correction signal may not exist for this problem.

Good work thorfinn — the clean implementation and thorough analysis is noted. Moving you to a new hypothesis..
